File: Labs/Lab 6/canFinish.java
import java.util.ArrayList;

public class canFinish {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        ArrayList<Integer>[] graph = new ArrayList[numCourses];
        for (int i=0;i<numCourses;i++) {
            graph[i] = new ArrayList<>();
        }
        
        for (int[] prerequisite : prerequisites) {
            int course = prerequisite[0];
            int prerequisiteCourse = prerequisite[1];
            graph[prerequisiteCourse].add(course);
        }
        
        boolean[] visited = new boolean[numCourses];
        boolean[] recStack = new boolean[numCourses];
        
        class GraphCycleDetector {
            public boolean isCyclic(int course) {
                if (recStack[course]) {
                    return true;
                }
                if (visited[course]) {
                    return false;
                }
                
                visited[course] = true;
                recStack[course] = true;
                
                for (Integer neighbor : graph[course]) {
                    if (isCyclic(neighbor)) {
                        return true;
                    }
                }
                
                recStack[course] = false;
                return false;
            }
        }
        
        GraphCycleDetector g = new GraphCycleDetector();

        for (int i=0;i<numCourses;i++) {
            if (!visited[i]) { 
                if (g.isCyclic(i)) {
                    return false;
                }
            }
        }
        
        return true;
    }
}

File: Labs/Lab 6/canFinish.py
def canFinish(numCourses, prerequisites):
    graph = [[] for _ in range(numCourses)]
    for dest, src in prerequisites:
        graph[src].append(dest)
    
    visited = [False] * numCourses
    recStack = [False] * numCourses

    def dfs(course):
        if recStack[course]:
            return False
        if visited[course]:
            return True
        
        visited[course] = True
        recStack[course] = True

        for neighbor in graph[course]:
            if not dfs(neighbor):
                return False
        
        recStack[course] = False
        return True
    
    for course in range(numCourses):
        if not visited[course]:
            if not dfs(course):
                return False

    return True
